File: mrb_doc/audio/sound.rb
# Plays a sound object.
# @param sound [Sound]
# @return [nil]
def play_sound(sound)
  # mrb_play_sound
  # src/mruby_integration/audio/sound.cpp
  nil
end

# Stops a sound object.
# @param sound [Sound]
# @return [nil]
def stop_sound(sound)
  # mrb_stop_sound
  # src/mruby_integration/audio/sound.cpp
  nil
end

# Pauses a sound object so it can be resumed.
# @param sound [Sound]
# @return [nil]
def pause_sound(sound)
  # mrb_pause_sound
  # src/mruby_integration/audio/sound.cpp
  nil
end

# Resumes a paused sound object.
# @param sound [Sound]
# @return [nil]
def resume_sound(sound)
  # mrb_resume_sound
  # src/mruby_integration/audio/sound.cpp
  nil
end

# Checks whether or not a sound object is currently playing.
# @param sound [Sound]
# @return [Boolean]
def sound_playing?(sound)
  # mrb_sound_playing
  # src/mruby_integration/audio/sound.cpp
  true
end

# Sets the volume of the sound object.
# @param sound [Sound]
# @param volume [Float] A value between 0.0 and 1.0.
# @return [nil]
def set_sound_volume(sound, volume)
  # mrb_set_sound_volume
  # src/mruby_integration/audio/sound.cpp
  1
end

# Sets the pitch of the sound object.
# @param sound [Sound]
# @param pitch [Float]
# @return [nil]
def set_sound_pitch(sound, pitch)
  # mrb_set_sound_pitch
  # src/mruby_integration/audio/sound.cpp
  1
end

File: src/mruby_integration/audio/sound.cpp
#include "mruby.h"
#include "raylib.h"
#include <cstdlib>

#include "mruby_integration/models/sound.hpp"
#include "mruby_integration/struct_types.hpp"

auto
mrb_pause_sound(mrb_state* mrb, mrb_value) -> mrb_value
{
  Sound* sound;
  mrb_get_args(mrb, "d", &sound, &Sound_type);

  PauseSound(*sound);

  return mrb_nil_value();
}

auto
mrb_resume_sound(mrb_state* mrb, mrb_value) -> mrb_value
{
  Sound* sound;
  mrb_get_args(mrb, "d", &sound, &Sound_type);

  ResumeSound(*sound);

  return mrb_nil_value();
}

auto
mrb_sound_playing(mrb_state* mrb, mrb_value) -> mrb_value
{
  Sound* sound;
  mrb_get_args(mrb, "d", &sound, &Sound_type);

  return mrb_bool_value(IsSoundPlaying(*sound));
}

auto
mrb_set_sound_volume(mrb_state* mrb, mrb_value) -> mrb_value
{
  Sound* sound;
  mrb_float volume;
  mrb_get_args(mrb, "df", &sound, &Sound_type, &volume);

  SetSoundVolume(*sound, volume);

  return mrb_nil_value();
}

auto
mrb_set_sound_pitch(mrb_state* mrb, mrb_value) -> mrb_value
{
  Sound* sound;
  mrb_float pitch;
  mrb_get_args(mrb, "df", &sound, &Sound_type, &pitch);

  SetSoundPitch(*sound, pitch);

  return mrb_nil_value();
}

void
append_audio_sound(mrb_state* mrb)
{
  mrb_define_method(
    mrb, mrb->kernel_module, "pause_sound", mrb_pause_sound, MRB_ARGS_REQ(1));
  mrb_define_method(
    mrb, mrb->kernel_module, "resume_sound", mrb_resume_sound, MRB_ARGS_REQ(1));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "sound_playing?",
                    mrb_sound_playing,
                    MRB_ARGS_REQ(1));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "set_sound_volume",
                    mrb_set_sound_volume,
                    MRB_ARGS_REQ(2));
  mrb_define_method(mrb,
                    mrb->kernel_module,
                    "set_sound_pitch",
                    mrb_set_sound_pitch,
                    MRB_ARGS_REQ(2));
}
